Remove __pycache__ when package is removed

vfazio opened the pull request:
Related to #11835

This PR has been updated to completely remove the adjacent `__pycache__` when a file is selected for removal during package uninstall.

If we find there are issues with not be selective when nuking this directory (due to maybe nuking pyc for packages sharing the same namespace), we can revert to a more conservative approach of including pyc files for each possible optimization level (the initial draft of this PR went this route), but that will not cover scenarios of remnant pyc files having been generated by multiple interpreter versions (as we will only uninstall for the currently active interpreter version and the version used during installation if bytecode was compiled as part of installation).

We pivoted to this method because `uv` seems to nuke the entire `__pycache__` directory with little ill effect. 

~~This should partially address the scenario where a module has been loaded by an interpreter with various optimization levels and has generated multiple pyc files.~~

~~This does not resolve scenarios where pyc files have been generated by multiple versions of an interpreter or multiple interpreter implementations.~~

~~For scenarios where there are files for multiple versions or interpreters, explicit virtual environments may be a more correct solution since dependency trees change based on interpreter and version.~~

~~Otherwise, short of nuking all of `__pycache__` which may affect cached bytecode for packages sharing a namespace or moving it out of the package install directory, this seems like a reasonable stop gap for this specific situation.~~


<!---
Thank you for your soon to be pull request. Before you submit this, please
double check to make sure that you've added a news file fragment. In pip we
generate our NEWS.rst from multiple news fragment files, and all pull requests
require either a news file fragment or a marker to indicate they don't require
one.

To read more about adding a news file fragment for your PR, please check out
our documentation at: https://pip.pypa.io/en/latest/development/contributing/#news-entries
-->

vfazio committed:
Remove __pycache__ directories during package removal

Previously, pip would only remove .pyc files from __pycache__ based on
entries that may have appeared in the RECORD file due to pip compiling
bytecode as part of package installation and based on the optimization
level of the active interpreter due to how `cache_from_source` works.

PEP-488 [0] mentions that different interpreter optimization levels may
generate unique .pyc files in __pycache__ so a module loaded by an
interpreter running at an optimization level other than what pip is run
with during installation and/or package removal may cause files to be
orphaned in __pycache__.

This can cause problems since orphaned files in __pycache__ directories
can make a package look like a PEP-420 importable namespace despite
having no runnable source files or "sourceless" .pyc files [1]. Code
trying to use `import <module>` to test for the presence of a package
may inadvertently pass in these situations.

Now, whenever a .py file is selected for removal as part of a package
being uninstalled, the adjacent __pycache__ directory is invalidated and
all of its contents are selected for removal (regardless of extension).

Note: any modules sharing a namespace with a package that was removed
will have their .pyc files removed as well, however, the .pyc files will
be regenerated the next time the interpreter loads the module.

[0]: https://peps.python.org/pep-0488/#proposal
[1]: https://peps.python.org/pep-3147/#case-4-legacy-pyc-files-and-source-less-imports

Signed-off-by: Vincent Fazio <vfazio@gmail.com>

vfazio commented on src/pip/_internal/req/req_uninstall.py:
```diff
@@ -336,9 +336,13 @@ def add(self, path: str) -> None:
+            os.path.splitext(path)[1] == ".py"
+            and sys.implementation.cache_tag is not None
+        ):
+            for opt in ("", 1, 2):
```
I did note uv's behavior in https://github.com/pypa/pip/issues/11835#issuecomment-3697125546

It does seem problematic that:

ns-package-1: `a/b/file.py`
ns-package-2: `a/b/file2.py`

will both utilize `a/b/__pycache__` and the optimized files will be nuked if if either of those packages get reinstalled. Subsequent module loading should recompile the bytecode however.

This change seemed like the less intrusive option. I don't know if other tools utilize `__pycache__`. I think there was some discussion about that in the linked issue.

Since `pip` is utilized by a lot of people, I wanted to take a conservative approach here. But if the maintainers feel it's appropriate to nuke `__pycache__` wholesale, I can change the code.

I will update the commit to link to some documentation about the optimization levels.

vfazio replied:
It suppose it could be argued that anything in `__pycache__` is null and void if the package is uninstalled, especially if the expectation is that the interpreter or other mechanism can just regenerate the optimized/cached files. 

`__pycache__` should only be generated for .py files without a side-by-side .pyc I believe, so the removal of any .py specified in a RECORD should maybe signal that `__pycache__`, if it exists, should be removed?

Open to thoughts

vfazio commented:
@notatallshaw No problems at all. Heck, I only have time to do this because it's the end of the year and i have some spare cycles myself so no rush on my part, really.

I think the biggest thing is deciding how we want pip to react in this scenario. This PR is pretty focused on a single known case, but if we wanted to mirror how `uv` is doing things and just blow away `__pycache__` when a package is removed with the expectation that the interpreter or other mechanism regenerates files as necessary, I think that may be reasonable too but does have more implications for consumers of the files in that directory (which shouldn't be reliable after a package has been removed).

vfazio commented:
Doing a blind push for `__pycache__` removal since I can't get the tests to run locally

notatallshaw commented:
@vfazio I do want to carefully review and consider this PR, and the different approaches, but I am not going to have time before 26.0 is released at this end of this month. So for me the earliest I'll get chance some time in February. Apologies for the slow pace, but just the situation we're in with the few active maintainers only able to donate their spare time.

vfazio commented:
> @vfazio I do want to carefully review and consider this PR, and the different approaches, but I am not going to have time before 26.0 is released at this end of this month. So for me the earliest I'll get chance some time in February. Apologies for the slow pace, but just the situation we're in with the few active maintainers only able to donate their spare time.

:100: understand. I just want to make sure if there's something i need to provide that I do so and to subtly bump the PR. I've had a few PRs go stale simply because i wasn't proactive enough in bumping the changes, so i try to nudge up until i'm just shy of being annoying

vfazio commented:
> And on the elif branch in add() (req_uninstall.py) the files added via os.walk in the elif branch bypass the normalization, existence check, and _permitted check that add() normally performs, perhaps recursively calling self.add() for each file instead of directly updating self._paths? e.g.
> 
> ```python
>   elif os.path.basename(path) == "__pycache__" and os.path.isdir(path):                                                                                                                                                                                                                             
>       for dirpath, _, dirfiles in os.walk(path):                                                                                                                                                                                                                                                    
>           for f in dirfiles:                                                                                                                                                                                                                                                                        
>               self.add(os.path.join(dirpath, f))                                                                                                                                                                                                                                                    
> ```
> 
> I think that makes sense, but perhaps I'm missing something?

I think I was trying to avoid that because `os.walk` should only report files that exist in path, the path itself has already been normalized, and we know the path is permitted.

The performance benefit is likely negligible; avoiding a `stat` call and some string manipulation, so recursively calling is probably fine.

vfazio commented:
Reworded commit slightly

pfmoore commented on src/pip/_internal/req/req_uninstall.py:
```diff
@@ -336,9 +335,16 @@ def add(self, path: str) -> None:
 
         # __pycache__ files can show up after 'installed-files.txt' is created,
-        # due to imports
+        # due to imports and is invalidated when an adjacent .py file is removed
```
I'm not sure what the added text means here. Could you reword it, please?

vfazio replied:
I've expounded on this a bit, lmk if it's too verbose or still not clear. I tried to also capture most of the rationale in the commit message itself.

pfmoore commented on src/pip/_internal/req/req_uninstall.py:
```diff
@@ -336,9 +335,16 @@ def add(self, path: str) -> None:
         if os.path.splitext(path)[1] == ".py":
-            self.add(cache_from_source(path))
+            pycache = os.path.join(os.path.dirname(path), "__pycache__")
+            # only evaluate __pycache__ once
```
Maybe expand on this comment, as it's not immediately obvious to me why you'd recursively call `add` on the `__pycache__` directory, rather than simply inlining the `elif` branch just below (it's not as if callers will be explicitly adding `__pycache__` anywhere else).

I assume it's something to do with the `_permitted` check, but I don't fully understand why. To be fair, the comments for `_permitted` don't give much indication of what it's for (and the implementation doesn't shed much light on it either!) so you shouldn't worry *too* much here. But some indication of your reasoning would help future maintainers.

vfazio replied:
@pfmoore thanks for taking a look. In light of your other comment, let me do some quick testing. The original idea was if multiple files in the same namespace get evaluated, the adjacent `__pycache__` directory only needs to be evaluated once.

Probably similar to https://github.com/pypa/pip/pull/13725#issuecomment-4017741672 this is likely a micro-optimization to avoid checks that we know should succeed and needless calls to `stat`.

pfmoore commented on src/pip/_internal/req/req_uninstall.py:
```diff
@@ -336,9 +335,16 @@ def add(self, path: str) -> None:
+            # only evaluate __pycache__ once
+            if not (pycache in self._paths or pycache in self._refuse):
+                self.add(pycache)
+        elif os.path.basename(path) == "__pycache__" and os.path.isdir(path):
```
Actually, why are we even doing a path walk here? We've just added `__pycache__` to the `_paths` set, and the code which processes that set is capable of handling directories, so why do anything special here?

vfazio replied:
@pfmoore let me test that real quick... I feel like i ran into some issue with that, like maybe the summary of the removed files wasn't rolled up or something, but I may be misremembering from when this code was originally adding pyc files explicitly and expecting the code that compresses the list of files for removal to roll up the directory

vfazio replied:
hmm... Seems to work fine just adding the directory...

I'll push these changes

```
vfazio@Zephyrus:/tmp/tmp.aTBZBg7bE1$ python3 -m venv .venv
vfazio@Zephyrus:/tmp/tmp.aTBZBg7bE1$ . .venv/bin/activate
(.venv) vfazio@Zephyrus:/tmp/tmp.aTBZBg7bE1$ pip install ~/development/pip/dist/pip-26.1.dev0-py3-none-any.whl
Processing /home/vfazio/development/pip/dist/pip-26.1.dev0-py3-none-any.whl
Installing collected packages: pip
  Attempting uninstall: pip
    Found existing installation: pip 24.0
    Uninstalling pip-24.0:
      Successfully uninstalled pip-24.0
Successfully installed pip-26.1.dev0
(.venv) vfazio@Zephyrus:/tmp/tmp.aTBZBg7bE1$ PYTHONOPTIMIZE=2 pip install anyio==4.12.0
Collecting anyio==4.12.0
  Downloading anyio-4.12.0-py3-none-any.whl.metadata (4.3 kB)
Collecting idna>=2.8 (from anyio==4.12.0)
  Downloading idna-3.11-py3-none-any.whl.metadata (8.4 kB)
Collecting typing_extensions>=4.5 (from anyio==4.12.0)
  Using cached typing_extensions-4.15.0-py3-none-any.whl.metadata (3.3 kB)
Downloading anyio-4.12.0-py3-none-any.whl (113 kB)
Downloading idna-3.11-py3-none-any.whl (71 kB)
Using cached typing_extensions-4.15.0-py3-none-any.whl (44 kB)
Installing collected packages: typing_extensions, idna, anyio
Successfully installed anyio-4.12.0 idna-3.11 typing_extensions-4.15.0

(.venv) vfazio@Zephyrus:/tmp/tmp.aTBZBg7bE1$ ls .venv/lib/python3.12/site-packages/anyio/__pycache__/
__init__.cpython-312.opt-2.pyc        lowlevel.cpython-312.opt-2.pyc        to_process.cpython-312.opt-2.pyc
from_thread.cpython-312.opt-2.pyc     pytest_plugin.cpython-312.opt-2.pyc   to_thread.cpython-312.opt-2.pyc
functools.cpython-312.opt-2.pyc       to_interpreter.cpython-312.opt-2.pyc
(.venv) vfazio@Zephyrus:/tmp/tmp.aTBZBg7bE1$ pip uninstall anyio
Found existing installation: anyio 4.12.0
Uninstalling anyio-4.12.0:
  Would remove:
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio-4.12.0.dist-info/*
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/*
Proceed (Y/n)? n
(.venv) vfazio@Zephyrus:/tmp/tmp.aTBZBg7bE1$ python3 -c "import anyio"

(.venv) vfazio@Zephyrus:/tmp/tmp.aTBZBg7bE1$ ls .venv/lib/python3.12/site-packages/anyio/__pycache__/
__init__.cpython-312.opt-2.pyc     functools.cpython-312.opt-2.pyc      to_interpreter.cpython-312.opt-2.pyc
__init__.cpython-312.pyc           lowlevel.cpython-312.opt-2.pyc       to_process.cpython-312.opt-2.pyc
from_thread.cpython-312.opt-2.pyc  lowlevel.cpython-312.pyc             to_thread.cpython-312.opt-2.pyc
from_thread.cpython-312.pyc        pytest_plugin.cpython-312.opt-2.pyc  to_thread.cpython-312.pyc
(.venv) vfazio@Zephyrus:/tmp/tmp.aTBZBg7bE1$ pip uninstall anyio
Found existing installation: anyio 4.12.0
Uninstalling anyio-4.12.0:
  Would remove:
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio-4.12.0.dist-info/*
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/*
Proceed (Y/n)? n

(.venv) vfazio@Zephyrus:/tmp/tmp.aTBZBg7bE1$ grep -A10 -a10 "pycache =" .venv/lib/python3.12/site-packages/pip/_internal/req/req_uninsta
ll.py
            self._refuse.add(path)

        # __pycache__ files can show up after 'installed-files.txt' is created,
        # due to imports.
        # Add the entire __pycache__ directory to the UninstallPathSet when an
        # adjacent .py file is removed to avoid the risk of orphaned .pyc files
        # created by a different interpreter version than the one running pip
        # at the time of uninstallation or an interpreter run at a different
        # optimization level.
        if os.path.splitext(path)[1] == ".py":
            pycache = os.path.join(os.path.dirname(path), "__pycache__")
            self.add(pycache)
        # elif os.path.basename(path) == "__pycache__" and os.path.isdir(path):
        #     for dirpath, _, dirfiles in os.walk(path):
        #         for f in dirfiles:
        #             self.add(os.path.join(dirpath, f))

    def add_pth(self, pth_file: str, entry: str) -> None:
        pth_file = self._normalize_path_cached(pth_file)
        if self._permitted(pth_file):
            if pth_file not in self._pth:
```

vfazio replied:
Yea, i think it was just that all of the pyc files are not listed in the verbose output of uninstall:

See below:

<details>

<summary>Verbose output</summary>

```
(.venv) vfazio@Zephyrus:/tmp/tmp.aTBZBg7bE1$ pip -vvv uninstall anyio
Found existing installation: anyio 4.11.0
Uninstalling anyio-4.11.0:
  Would remove:
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio-4.11.0.dist-info/INSTALLER
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio-4.11.0.dist-info/METADATA
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio-4.11.0.dist-info/RECORD
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio-4.11.0.dist-info/REQUESTED
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio-4.11.0.dist-info/WHEEL
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio-4.11.0.dist-info/entry_points.txt
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio-4.11.0.dist-info/licenses/LICENSE
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio-4.11.0.dist-info/top_level.txt
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/__init__.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/__pycache__
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_backends/__init__.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_backends/__pycache__
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_backends/_asyncio.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_backends/_trio.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/__init__.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/__pycache__
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/_asyncio_selector_thread.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/_contextmanagers.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/_eventloop.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/_exceptions.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/_fileio.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/_resources.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/_signals.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/_sockets.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/_streams.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/_subprocesses.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/_synchronization.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/_tasks.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/_tempfile.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/_testing.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/_typedattr.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/abc/__init__.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/abc/__pycache__
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/abc/_eventloop.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/abc/_resources.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/abc/_sockets.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/abc/_streams.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/abc/_subprocesses.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/abc/_tasks.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/abc/_testing.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/from_thread.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/lowlevel.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/py.typed
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/pytest_plugin.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/streams/__init__.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/streams/__pycache__
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/streams/buffered.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/streams/file.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/streams/memory.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/streams/stapled.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/streams/text.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/streams/tls.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/to_interpreter.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/to_process.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/to_thread.py
  Will actually move:
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio-4.11.0.dist-info/
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/__init__.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/__pycache__
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_backends/
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/__init__.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/__pycache__
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/_asyncio_selector_thread.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/_contextmanagers.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/_eventloop.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/_exceptions.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/_fileio.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/_resources.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/_signals.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/_sockets.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/_streams.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/_subprocesses.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/_synchronization.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/_tasks.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/_tempfile.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/_testing.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/_core/_typedattr.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/abc/__init__.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/abc/__pycache__
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/abc/_eventloop.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/abc/_resources.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/abc/_sockets.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/abc/_streams.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/abc/_subprocesses.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/abc/_tasks.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/abc/_testing.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/from_thread.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/lowlevel.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/py.typed
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/pytest_plugin.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/streams/__init__.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/streams/__pycache__
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/streams/buffered.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/streams/file.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/streams/memory.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/streams/stapled.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/streams/text.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/streams/tls.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/to_interpreter.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/to_process.py
    /tmp/tmp.aTBZBg7bE1/.venv/lib/python3.12/site-packages/anyio/to_thread.py
```

</details>

If this is OK, i can keep this change

pfmoore replied:
I'm pretty sure we *don't* want all of the `.pyc` files listed individually, so IMO this is actually better.

vfazio commented:
I derped the unit test with this change, so need to fix that rq

vfazio commented:
> I derped the unit test with this change, so need to fix that rq

I hacked this real quick just to get something to pass. I'm not sure this should be the final form of the unit test because it doesn't really reflect what we'd expect to see when a RECORD gets parsed to queue paths for removal I don't think (but haven't verified).

In the case of a RECORD, the pyc files maybe be in the set of paths... but maybe this is a good enough mockup to verify things are working as expected.

pfmoore reviewed (approved):
Looks good, thanks!

I feel like the unit test is being a little too detailed in asserting that the `.pyc` file isn't in `_paths`, but it's the nature of unit tests to be low-level, so I think it's probably OK as it is.

vfazio commented:
> Looks good, thanks!
> 
> I feel like the unit test is being a little too detailed in asserting that the `.pyc` file isn't in `_paths`, but it's the nature of unit tests to be low-level, so I think it's probably OK as it is.

I can remove those two lines in a couple of hours when I have access to my laptop again. Just lmk if that's preferred.

pfmoore commented:
I'm fine with leaving them, unless other maintainers think they should go.

notatallshaw commented:
One small thing I spotted that would be good to clean up before merge, but isn't a showstopper, this PR adds the `__pycache__` dir to `_paths`, but its contents are not added individually. `compress_for_output_listing()` only checks individual paths, so a stray file like `__pycache__/HELLO` gets listed under "Would not remove", but it actually does get removed.

A fix could be in In `compress_for_output_listing()`, skip adding a file to `will_skip` if any ancestor directory is already in `paths`.

vfazio commented:
> One small thing I spotted that would be good to clean up before merge, but isn't a showstopper, this PR adds the `__pycache__` dir to `_paths`, but its contents are not added individually. `compress_for_output_listing()` only checks individual paths, so a stray file like `__pycache__/HELLO` gets listed under "Would not remove", but it actually does get removed.
> 
> A fix could be in In `compress_for_output_listing()`, skip adding a file to `will_skip` if any ancestor directory is already in `paths`.

I have maybe too many questions which is why I'm probably not well suited to be doing development at this level, but i'll ask anyway:

* Does this mean `test_compressed_listing` is not behaving as expected currently even without my changes? 
* Is the `path` argument to `UninstallPathSet.add` supposed to be an individual file path (`is_file`) and not a folder (meaning that adding `__pycache__` is not really supported?
  * If `path` can be a directory, should `compress_for_output_listing` be performing an `is_dir` check to add these paths to the set of folders? I don't know if there are weird situations we need to be aware of like symlinks pointing to directories and needing to use realpath?

Searching through all `paths` for each entry in `paths` is quadratic in the worst case scenario (I think?), so I assume we should avoid that. If we can reduce the set to be just the `folders` (where adding an `is_dir` check could populate this) may help shave the time complexity down. But this only makes sense if the `UninstallPathSet.add` function is supposed to allow directories and not.

If it's easier to just walk the directory and add the files individually, I can revert my recent changes. They were done based on feedback so just let me know everyone's preference.

vfazio commented:
So, i have this diff, but I kinda feel like i'm fighting with `compress_for_rename` so i'm having to add a trailing slash to path names:

```diff
(.venv) vfazio@Zephyrus:~/development/pip$ PAGER= git diff
diff --git a/src/pip/_internal/req/req_uninstall.py b/src/pip/_internal/req/req_uninstall.py
index adb215c35..1532819e1 100644
--- a/src/pip/_internal/req/req_uninstall.py
+++ b/src/pip/_internal/req/req_uninstall.py
@@ -2,6 +2,7 @@ from __future__ import annotations
 
 import functools
 import os
+import pathlib
 import sys
 import sysconfig
 from collections.abc import Generator, Iterable
@@ -157,7 +158,10 @@ def compress_for_output_listing(paths: Iterable[str]) -> tuple[set[str], set[str
     # Determine folders and files
     folders = set()
     files = set()
+    selected_directories = set()
     for path in will_remove:
+        if os.path.isdir(path):
+            selected_directories.add(path)
         if path.endswith(".pyc"):
             continue
         if path.endswith("__init__.py") or ".dist-info" in path:
@@ -172,6 +176,10 @@ def compress_for_output_listing(paths: Iterable[str]) -> tuple[set[str], set[str
     # that might get added.
     for folder in folders:
         for dirpath, _, dirfiles in os.walk(folder):
+            if any(
+                pathlib.Path(dirpath).is_relative_to(d) for d in selected_directories
+            ):
+                continue
             for fname in dirfiles:
                 if fname.endswith(".pyc"):
                     continue
@@ -342,7 +350,7 @@ class UninstallPathSet:
         # pip at the time of package installation and uninstallation or an
         # interpreter run at a different optimization level (PYTHONOPTIMIZE).
         if os.path.splitext(path)[1] == ".py":
-            pycache = os.path.join(os.path.dirname(path), "__pycache__")
+            pycache = os.path.join(os.path.dirname(path), "__pycache__" + os.sep)
             self.add(pycache)
 
     def add_pth(self, pth_file: str, entry: str) -> None:
diff --git a/tests/unit/test_req_uninstall.py b/tests/unit/test_req_uninstall.py
index abb5fa2d0..49d1d465c 100644
--- a/tests/unit/test_req_uninstall.py
+++ b/tests/unit/test_req_uninstall.py
@@ -72,6 +72,7 @@ def test_compressed_listing(tmpdir: Path) -> None:
             "lib/mypkg/__init__.py",
             "lib/mypkg/my_awesome_code.py",
             "lib/mypkg/__pycache__/my_awesome_code-magic.pyc",
+            "lib/mypkg/__pycache__/.skip.garbage",
             "lib/mypkg/support/support_file.py",
             "lib/mypkg/support/more_support.py",
             "lib/mypkg/support/would_be_skipped.skip.py",
@@ -87,6 +88,14 @@ def test_compressed_listing(tmpdir: Path) -> None:
 
     # Remove the files to be skipped from the paths
     sample = [path for path in sample if ".skip." not in path]
+    sample.extend(
+        in_tmpdir(
+            [
+                "lib/mypkg/__pycache__/",
+                "lib/mypkg/support/__pycache__/",
+            ]
+        )
+    )
 
     expected_remove = in_tmpdir(
         [
@@ -249,7 +258,7 @@ class TestUninstallPathSet:
         )
 
         pkg_dir = os.path.normcase(os.path.join(tmpdir, "mypkg"))
-        pycache_dir = os.path.join(pkg_dir, "__pycache__")
+        pycache_dir = os.path.join(pkg_dir, "__pycache__", "")
         os.makedirs(pycache_dir)
 
         py_file = os.path.join(pkg_dir, "foo.py")
@@ -269,8 +278,6 @@ class TestUninstallPathSet:
         # we should not expect to see the pyc files explicitly
         assert py_file in ups._paths
         assert pycache_dir in ups._paths
-        assert pyc_file not in ups._paths
-        assert opt_pyc_file not in ups._paths
 
 
 class TestStashedUninstallPathSet:
```

vfazio commented:
Maybe this hassle is a sign that `path` is supposed to be individual file paths and not a directory? I'll have more time in a few hours maybe to look into this,

notatallshaw commented:
The fault is mine, I didn't fully engage with the full PR (I'm trying to crank out multiple reviews this evening).

> Does this mean test_compressed_listing is not behaving as expected currently even without my changes?

`test_compressed_listing` on `main` isn't affected, its sample has individual `.pyc` paths, never a `__pycache__` directory.

> Is the path argument to UninstallPathSet.add supposed to be an individual file path (is_file) and not a folder (meaning that adding __pycache__ is not really supported?

`add()` already accepts directories, as `stash()` branches on `os.path.isdir`. It also guards symlinks (`isdir and not islink`), so a symlinked `__pycache__` is stashed as a file. No need to revert to walking `__pycache__` and adding files individually in `add()`.

> If path can be a directory, should compress_for_output_listing be performing an is_dir check to add these paths to the set of folders?

Don't merge these into `folders`. That set means "package roots to walk"; putting `__pycache__` there would make the walker descend into it and emit wrong output. A separate set avoids touching the existing logic.

> I don't know if there are weird situations we need to be aware of like symlinks pointing to directories and needing to use realpath?

No extra `realpath` needed, `add()` already runs `normalize_path` (which calls `realpath`) on the directory part, so `_paths` and the paths `os.walk` yields are already canonical.

> Searching through all paths for each entry in paths is quadratic in the worst case scenario (I think?), so I assume we should avoid that.

Yeah, my suggestion was quadratic, you can instead prune `os.walk` via `subdirs[:]`:

```python
_scheduled_dirs = {
    os.path.normcase(p) for p in will_remove
    if os.path.isdir(p) and not os.path.islink(p)
}

for folder in folders:
    for dirpath, subdirs, dirfiles in os.walk(folder):
        subdirs[:] = [
            d for d in subdirs
            if os.path.normcase(os.path.join(dirpath, d))
            not in _scheduled_dirs
        ]
        for fname in dirfiles:
            ...  # existing body unchanged
```

Like I said though, this is minor it's edge case behavior and currently untested, we can handle this in a follow up PR if it's needed.

Your diff looks close but you could use subdirectory pruning and you shouldn't be manually handling `os.sep`, pathlib.Path normalizes trailing slashes and os.path.isdir handles them either way.

vfazio commented:
Late night drivel -

Might be worth a new issue then... `add` may accept paths and `stash` handles the paths returned from `compress_for_rename` but that function performs a split on the path that drops the directory name because there is no trailing slash

```python
def compress_for_rename(paths: Iterable[str]) -> set[str]:
    """Returns a set containing the paths that need to be renamed.

    This set may include directories when the original sequence of paths
    included every file on disk.
    """
    case_map = {os.path.normcase(p): p for p in paths}
    remaining = set(case_map)
    unchecked = sorted({os.path.split(p)[0] for p in case_map.values()}, key=len)  # the final path component is dropped since directories have no trailing slash in `paths`
```

It seems then the directory suffix is treated as a file since its parent is used in the loop.

Also, `compress_for_rename` returns wildcards as ending with a `os.path.sep` and I think `compact` uses that as a marker for collapsing paths.

My quick and unoptimized solution is:

```diff
(.venv) vfazio@Zephyrus:~/development/pip$ PAGER= git diff
diff --git a/src/pip/_internal/req/req_uninstall.py b/src/pip/_internal/req/req_uninstall.py
index adb215c35..9bdc4afcc 100644
--- a/src/pip/_internal/req/req_uninstall.py
+++ b/src/pip/_internal/req/req_uninstall.py
@@ -114,12 +114,31 @@ def compress_for_rename(paths: Iterable[str]) -> set[str]:
     """
     case_map = {os.path.normcase(p): p for p in paths}
     remaining = set(case_map)
-    unchecked = sorted({os.path.split(p)[0] for p in case_map.values()}, key=len)
     wildcards: set[str] = set()
 
     def norm_join(*a: str) -> str:
         return os.path.normcase(os.path.join(*a))
 
+    # select all files in directories scheduled for removal as a first sweep
+    _dirs = {
+        key: path
+        for key, path in case_map.items()
+        if os.path.isdir(path) and not os.path.islink(path)
+    }
+
+    for key, path in _dirs.items():
+        _files: set[str] = set()
+        for dirname, _, files in os.walk(path):
+            _files.update(norm_join(path, dirname, f) for f in files)
+        remaining.difference_update(_files)
+        # remove the directory from the list of remaining items to process
+        # and add it to the list of directory wildcards
+        remaining.remove(key)
+        case_map.pop(key, None)
+        wildcards.add(path + os.sep)
+
+    unchecked = sorted({os.path.split(p)[0] for p in case_map.values()}, key=len)
+
     for root in unchecked:
         if any(os.path.normcase(root).startswith(w) for w in wildcards):
             # This directory has already been handled.
@@ -157,7 +176,12 @@ def compress_for_output_listing(paths: Iterable[str]) -> tuple[set[str], set[str
     # Determine folders and files
     folders = set()
     files = set()
+
+    _scheduled_dirs = set()
+
     for path in will_remove:
+        if os.path.isdir(path) and not os.path.islink(path):
+            _scheduled_dirs.add(os.path.normcase(path))
         if path.endswith(".pyc"):
             continue
         if path.endswith("__init__.py") or ".dist-info" in path:
@@ -171,7 +195,13 @@ def compress_for_output_listing(paths: Iterable[str]) -> tuple[set[str], set[str
     # This walks the tree using os.walk to not miss extra folders
     # that might get added.
     for folder in folders:
-        for dirpath, _, dirfiles in os.walk(folder):
+        for dirpath, subdirs, dirfiles in os.walk(folder):
+            subdirs[:] = [
+                d
+                for d in subdirs
+                if os.path.normcase(os.path.join(dirpath, d)) not in _scheduled_dirs
+            ]
+
             for fname in dirfiles:
                 if fname.endswith(".pyc"):
                     continue
diff --git a/tests/unit/test_req_uninstall.py b/tests/unit/test_req_uninstall.py
index abb5fa2d0..f9e9654ed 100644
--- a/tests/unit/test_req_uninstall.py
+++ b/tests/unit/test_req_uninstall.py
@@ -72,6 +72,7 @@ def test_compressed_listing(tmpdir: Path) -> None:
             "lib/mypkg/__init__.py",
             "lib/mypkg/my_awesome_code.py",
             "lib/mypkg/__pycache__/my_awesome_code-magic.pyc",
+            "lib/mypkg/__pycache__/.skip.garbage",
             "lib/mypkg/support/support_file.py",
             "lib/mypkg/support/more_support.py",
             "lib/mypkg/support/would_be_skipped.skip.py",
@@ -87,6 +88,14 @@ def test_compressed_listing(tmpdir: Path) -> None:
 
     # Remove the files to be skipped from the paths
     sample = [path for path in sample if ".skip." not in path]
+    sample.extend(
+        in_tmpdir(
+            [
+                "lib/mypkg/__pycache__",
+                "lib/mypkg/support/__pycache__",
+            ]
+        )
+    )
 
     expected_remove = in_tmpdir(
         [
```

notatallshaw commented:
Yeah, it's edge case enough and could do with a review of the logic that let's make that it's own issue/PR.

notatallshaw commented:
Thanks for your PR and time and patience sticking with it.

vfazio commented:
@notatallshaw @pfmoore thanks for taking the time to review this and engaging with an amateur!

vfazio commented:
Seems like I actually broke something with this https://github.com/pypa/virtualenv/pull/3132/changes/6bf67165d31c4cfb206a9a5e3f67bfe651b7af21
